### ISSUE TITLE
Cannot edit / delete certain works

### ISSUE THREAD

**Issue by hackmastera** [chemheritage/chf-sufia#323]
zg64tk92g
m039k4882
r494vk17h

**Comment by hackmastera** [comment]
edit permissions are as expected.

stack trace is here https://gist.github.com/sanfordd/862bfa4e7abfbe733013474ce851c174

**Comment by hackmastera** [comment]
The error occurs when an index event is triggered on the beckman collection.

**Comment by hackmastera** [comment]
object counts are off for beckman:

```
> c.member_ids.count
=> 1000
irb(main):039:0> c.ordered_member_ids.count
=> 958
```

**Comment by sanfordd** [comment]
Here's an error I saw cropping up in Friday's logs a few times that may be related.
https://gist.github.com/sanfordd/f71ae742c05e7bf8edcfbe0cfb51f78f

**Comment by sanfordd** [comment]
Here's another error that appeared once, but may still be related.
https://gist.github.com/sanfordd/6cb3129b3d6db20514b3fb0fbda7a8f6

**Comment by sanfordd** [comment]
Last but not least, I saw this once Friday:
F, [2017-02-03T11:41:07.823900 #22181] FATAL -- : 
SystemStackError (stack level too deep):
  app/views/curation_concerns/base/show.html.erb:16:in `_app_views_curation_concerns_base_show_html_erb___306509359968677659_75143500'

**Comment by hackmastera** [comment]
Okay the 1st new stack trace is the same as the original one.

The 2nd is one we have seen before, #317, which was caused by a work having inconsistent member lists. I believe the member lists were inconsistent due to the background jobs / edit form overwrite problem but I'm not 100% sure about that.

The last error is one we see when a work has a thumbnail image in the index, but that image is not available to the viewer.

@sanfordd do you have any sense of which error came first?

**Comment by hackmastera** [comment]
the problem is that there are `nil`s in the ordered_members list for the beckman collection. I'm not sure how they got there but I suspect that the collection edit form or add to collection activity was open in more than one tab at a time. the collection save does not have proper locking on it. Will file an upstream bug once I manage to delete `nil` objects from that list.

**Comment by hackmastera** [comment]
```
> c.member_ids - c.ordered_member_ids
=> ["ht24wj78m", "xp68kg29t", "d791sg31t", "gx41mh94c", "79407x24n", "s1784k813", "5q47rn72z", "pg15bd88s", "fj236208t", "n009w229r", "4f16c2849", "k643b117x", "41687h474", "3484zg88m", "zp38wc60z", "ft848q60n", "tx31qh68h", "m039k4882", "jh343s559", "g732d897c", "kk91fk52b", "3j333225q", "c534fn93r", "mp48sc763", "qn59q396v", "hm50tr726", "xk81jk370", "cc08hf617", "th83kz38r", "xk81jk49s", "t148fh232", "2z10wq33g", "np193934k", "0k225b156", "5m60qs062", "df65v795x", "bc386j324", "j9602073n", "37720c78r", "j96020624", "x920fw89s", "br86b361k", "pg15bd96g", "0g354f281"]
irb(main):055:0> c.ordered_member_ids - c.member_ids
=> [nil, "1v53jx18g"]
```

**Comment by hackmastera** [comment]
User wanted to turn child works into child filesets. To achieve this, went to delete the child works and add them as files.

Operating on 2 parent works.

The first had 4 files, 2 child works. For the first, uploaded the files. Then deleted the the child works one at a time. This was successful on viewing the parent work.
parent work:
- mp48sc763

The second work had 2 child works. For the second, not sure exactly which she tried to do first but she was unable to delete either child work and unable to add new files -- both attempts generated the 'generic error' screen. She then deleted the file from the child work, successfully.
parent work:
- m039k4882: ["Beckman Model IR-33 Spectrophotometer"]
child works:
- zg64tk92g: ["Beckman Model IR-33 Spectrophotometer, top view"]
- h989r3203: front view

Yesterday tried to create a new work with files. was unable to add files to this work; probably haven't tried without adding directly to collection.

**Comment by hackmastera** [comment]
Okay I figured this out. During migration the members of a collection were added to `ordered_members`. But the sufia code never does that; it only uses `members`. So when the pre-migration object was deleted, it wasn't cleaned up from the ordered_members list since the code has no expectation of that list being in use.

We need to delete ordered_members from all collections.

I also need to fix this upstream since I am the one who introduced this bug.
